pkg/services/ngalert/prom/convert.go

```diff
@@ -159,62 +182,58 @@ func (p *Converter) convertRule(orgID int64, namespaceUID, group string, rule Pr
 		OrgID:        orgID,
 		NamespaceUID: namespaceUID,
 		Title:        title,
-		Data:         []models.AlertQuery{queryNode},
-		Condition:    "A",
+		Data:         query,
+		Condition:    query[len(query)-1].RefID,
 		NoDataState:  p.cfg.NoDataState,
 		ExecErrState: p.cfg.ExecErrState,
 		Annotations:  rule.Annotations,
 		Labels:       labels,
 		For:          forInterval,
 		RuleGroup:    group,
+		IsPaused:     isPaused,
+		Record:       record,
 		Metadata: models.AlertRuleMetadata{
 			PrometheusStyleRule: &models.PrometheusStyleRule{
 				OriginalRuleDefinition: string(originalRuleDefinition),
 			},
 		},
 	}
 
-	if rule.Record != "" {
-		result.Record = &models.Record{
-			From:   "A",
-			Metric: rule.Record,
-		}
-		result.IsPaused = p.cfg.RecordingRules.IsPaused
-	} else {
-		result.IsPaused = p.cfg.AlertRules.IsPaused
-	}
-
 	return result, nil
 }
 
-func createAlertQueryNode(datasourceUID, datasourceType, expr string, fromTimeRange, evaluationOffset time.Duration) (models.AlertQuery, error) {
-	modelData := map[string]interface{}{
-		"datasource": map[string]interface{}{
-			"type": datasourceType,
-			"uid":  datasourceUID,
-		},
-		"expr":    expr,
-		"instant": true,
-		"range":   false,
-		"refId":   "A",
+// createQuery constructs the alert query nodes for a given Prometheus rule expression.
+// It returns a slice of AlertQuery that represent the evaluation steps for the rule.
+//
+// For recording rules it generates a single query node that
+// executes the PromQL query in the configured datasource.
+//
+// For alerting rules, it generates three query nodes:
+//  1. Query Node (query): Executes the PromQL query using the configured datasource.
+//  2. Math Node (prometheus_math): Applies a math expression "is_number($query) || is_nan($query) || is_inf($query)".
+//  3. Threshold Node (threshold): Gets the result from the math node and checks that it's greater than 0.
+//
+// This is needed to ensure that we keep the Prometheus behaviour, where any returned result
+// is considered alerting, and only when the query returns no data is the alert treated as normal.
+func (p *Converter) createQuery(expr string, isRecordingRule bool) ([]models.AlertQuery, error) {
+	queryNode, err := createQueryNode(p.cfg.DatasourceUID, p.cfg.DatasourceType, expr, *p.cfg.FromTimeRange, *p.cfg.EvaluationOffset)
```

Comment:
I wonder if cfg can also contain `datasource.Datasource` rather than two fields. Perhaps, you will have to check if datasource exists (and user has permissions to it) while converting.